akka-actor/src/main/scala/akka/io/Inet.scala

```diff
@@ -16,38 +19,65 @@ object Inet {
     /**
      * Action to be taken for this option before bind() is called
      */
-    def beforeBind(ds: DatagramChannel): Unit = ()
+    def beforeDatagramBind(ds: DatagramSocket): Unit = ()
 
     /**
      * Action to be taken for this option before bind() is called
      */
-    def beforeBind(ss: ServerSocketChannel): Unit = ()
+    def beforeServerSocketBind(ss: ServerSocket): Unit = ()
 
     /**
-     * Action to be taken for this option before bind() is called
+     * Action to be taken for this option before calling connect()
      */
-    def beforeBind(s: SocketChannel): Unit = ()
+    def beforeConnect(s: Socket): Unit = ()
+
+    /**
+     * Action to be taken for this option after connect returned (i.e. on
+     * the slave socket for servers).
+     */
+    def afterConnect(s: Socket): Unit = ()
+  }
 
+  /**
+   * Java API: AbstractSocketOption is a package of data (from the user) and associated
+   * behavior (how to apply that to a channel).
+   */
+  abstract class AbstractSocketOption extends SocketOption
+
+  trait SocketOptionV2 extends SocketOption {
     /**
      * Action to be taken for this option after connect returned (i.e. on
      * the slave socket for servers).
      */
-    def afterConnect(c: DatagramChannel): Unit = ()
+    def afterBind(s: DatagramSocket): Unit = ()
 
     /**
      * Action to be taken for this option after connect returned (i.e. on
      * the slave socket for servers).
      */
-    def afterConnect(c: ServerSocketChannel): Unit = ()
+    def afterBind(s: ServerSocket): Unit = ()
 
     /**
      * Action to be taken for this option after connect returned (i.e. on
      * the slave socket for servers).
      */
-    def afterConnect(c: SocketChannel): Unit = ()
+    def afterConnect(s: DatagramSocket): Unit = ()
+
+    /**
+     * Action to be taken for this option after connect returned (i.e. on
+     * the slave socket for servers).
+     */
+    // FIXME where to call this one? Is it needed?
+    // def afterConnect(s: ServerSocket): Unit = ()
```

Comment:
AFAIU we already had it in the TCP case, which is why that is on SocketOption, and then we added UDP support later, which is now on SocketOptionV2. So I believe that everything should be okay now, what do you feel is missing?